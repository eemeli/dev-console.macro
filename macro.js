const { createMacro, MacroError } = require('babel-plugin-macros')
const plugin = require('babel-plugin-transform-remove-console')

const defaultConfig = {
  production: true,
  staging: { exclude: ['error'] }
}

function macro({ babel, config = defaultConfig, references, state }) {
  if (!references.default) {
    throw new MacroError('dev-console.macro requires a default import')
  }
  const opts = config[process.env.NODE_ENV]
  if (!opts) return

  // console needs to be in global scope for the plugin
  const { node, scope } = references.default[0]
  scope.removeBinding(node.name)
  scope.addGlobal(node)

  const { CallExpression, MemberExpression } = plugin(babel).visitor
  state.file.path.traverse({
    CallExpression(path) {
      CallExpression(path, { opts })
    },
    MemberExpression: {
      exit(path) {
        MemberExpression.exit(path, { opts })
      }
    }
  })
}

module.exports = createMacro(macro, { configName: 'remove-console' })
