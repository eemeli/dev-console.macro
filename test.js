import pluginTester from 'babel-plugin-tester'
import plugin from 'babel-plugin-macros'
import macro from './macro'

pluginTester({
  plugin,
  snapshot: true,
  babelOptions: { filename: __filename },
  tests: {
    basic: `
      import console from "./macro";
      function foo() {
        console.log("foo");
        blah();
      }
    `,
    guards: `
      import console from "./macro";
      function foo() {
        true && console.log("foo");
        blah();
      }
    `,
    'local-binding': `
      import console from "./macro";
      function foo(console) {
        console.foo("hi");
        const bar = console.foo.bind(console);
      }
      function bar(a) {
        const { console } = a;
        a.b = console => console.bar("bar");
        if (console.foo.call(console, "bar")) {
          return;
        }
      }
    `,
    'member-expr-assignment-no-op': `
      import console from "./macro";
      function foo() {
        console.foo = function foo() {
          console.log("foo");
        };
        console.error = myConsoleError;
        console.foo();
        console.error("asdf");
      }
    `,
    'member-expr-no-op': `
      import console from "./macro";
      const a = console.log;
      a();
      const b = console.log.bind(console);
      b("asdf");
      var x = console.log ? console.log('log') : foo();
      function foo() {
        if (console.error) {
          console.error("Errored");
        }
      }
      console.log.call(console, "foo");
      console.log.apply(null, {});
    `,
    'replace-with-empty-block': `
      import console from "./macro";
      if (blah) console.log(blah);
      for (;;) console.log(blah);
      for (var blah in []) console.log(blah);
      for (var blah of []) console.log(blah);
      while (blah) console.log(blah);
      do console.log(blah); while (blah);
    `,
    'top-level': `
      import console from "./macro";
      true && console.log("foo");
      blah();
    `,
    'top-level-stmts': `
      import console from "./macro";
      console.log("foo");
      blah();
    `
  }
})

describe('config', () => {
  let origOptions
  beforeAll(() => {
    origOptions = macro.options
    macro.options = { configName: 'test-remove-console' }
  })
  afterAll(() => {
    macro.options = origOptions
  })

  pluginTester({
    plugin,
    snapshot: true,
    babelOptions: { filename: __filename },
    tests: {
      'bound-excludes': `
        import console from "./macro";
        function foo() {
          const a = console.log;
          a();
          const b = console.error.bind(console);
          b("asdf");
          blah();
        }
      `,
      excludes: `
        import console from "./macro";
        function foo() {
          console.log("foo");
          console.error("bar");
          blah();
          console.info("blah");
        }
      `
    }
  })
})

describe('errors', () => {
  pluginTester({
    plugin,
    snapshot: false,
    babelOptions: { filename: __filename },
    tests: [
      {
        title: 'no default import',
        code: `
        import { foo } from "./macro";
          foo();
          blah();
        `,
        error: /requires a default import/
      },
      {
        title: 'bad default import name',
        code: `
        import foo from "./macro";
          foo();
          blah();
        `,
        error: /needs to have "console" as its local name/
      }
    ]
  })
})
