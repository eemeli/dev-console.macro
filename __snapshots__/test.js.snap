// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config macros bound-excludes: bound-excludes 1`] = `
"
import console from \\"./macro\\";
function foo() {
  const a = console.log;
  a();
  const b = console.error.bind(console);
  b(\\"asdf\\");
  blah();
}

      ↓ ↓ ↓ ↓ ↓ ↓

function foo() {
  const a = function () {};
  a();
  const b = console.error.bind(console);
  b(\\"asdf\\");
  blah();
}
"
`;

exports[`config macros excludes: excludes 1`] = `
"
import console from \\"./macro\\";
function foo() {
  console.log(\\"foo\\");
  console.error(\\"bar\\");
  blah();
  console.info(\\"blah\\");
}

      ↓ ↓ ↓ ↓ ↓ ↓

function foo() {
  console.error(\\"bar\\");
  blah();
  console.info(\\"blah\\");
}
"
`;

exports[`macros basic: basic 1`] = `
"
import console from \\"./macro\\";
function foo() {
  console.log(\\"foo\\");
  blah();
}

      ↓ ↓ ↓ ↓ ↓ ↓

function foo() {
  blah();
}
"
`;

exports[`macros guards: guards 1`] = `
"
import console from \\"./macro\\";
function foo() {
  true && console.log(\\"foo\\");
  blah();
}

      ↓ ↓ ↓ ↓ ↓ ↓

function foo() {
  true && void 0;
  blah();
}
"
`;

exports[`macros local-binding: local-binding 1`] = `
"
import console from \\"./macro\\";
function foo(console) {
  console.foo(\\"hi\\");
  const bar = console.foo.bind(console);
}
function bar(a) {
  const { console } = a;
  a.b = console => console.bar(\\"bar\\");
  if (console.foo.call(console, \\"bar\\")) {
    return;
  }
}

      ↓ ↓ ↓ ↓ ↓ ↓

function foo(console) {
  console.foo(\\"hi\\");
  const bar = console.foo.bind(console);
}
function bar(a) {
  const { console } = a;
  a.b = console => console.bar(\\"bar\\");
  if (console.foo.call(console, \\"bar\\")) {
    return;
  }
}
"
`;

exports[`macros member-expr-assignment-no-op: member-expr-assignment-no-op 1`] = `
"
import console from \\"./macro\\";
function foo() {
  console.foo = function foo() {
    console.log(\\"foo\\");
  };
  console.error = myConsoleError;
  console.foo();
  console.error(\\"asdf\\");
}

      ↓ ↓ ↓ ↓ ↓ ↓

function foo() {
  console.foo = function () {};
  console.error = function () {};
}
"
`;

exports[`macros member-expr-no-op: member-expr-no-op 1`] = `
"
import console from \\"./macro\\";
const a = console.log;
a();
const b = console.log.bind(console);
b(\\"asdf\\");
var x = console.log ? console.log('log') : foo();
function foo() {
  if (console.error) {
    console.error(\\"Errored\\");
  }
}
console.log.call(console, \\"foo\\");
console.log.apply(null, {});

      ↓ ↓ ↓ ↓ ↓ ↓

const a = function () {};
a();
const b = function () {};
b(\\"asdf\\");
var x = function () {} ? void 0 : foo();
function foo() {
  if (function () {}) {}
}
"
`;

exports[`macros replace-with-empty-block: replace-with-empty-block 1`] = `
"
import console from \\"./macro\\";
if (blah) console.log(blah);
for (;;) console.log(blah);
for (var blah in []) console.log(blah);
for (var blah of []) console.log(blah);
while (blah) console.log(blah);
do console.log(blah); while (blah);

      ↓ ↓ ↓ ↓ ↓ ↓

if (blah) {}
for (;;) {}
for (var blah in []) {}
for (var blah of []) {}
while (blah) {}
do {} while (blah);
"
`;

exports[`macros top-level: top-level 1`] = `
"
import console from \\"./macro\\";
true && console.log(\\"foo\\");
blah();

      ↓ ↓ ↓ ↓ ↓ ↓

true && void 0;
blah();
"
`;

exports[`macros top-level-stmts: top-level-stmts 1`] = `
"
import console from \\"./macro\\";
console.log(\\"foo\\");
blah();

      ↓ ↓ ↓ ↓ ↓ ↓

blah();
"
`;
